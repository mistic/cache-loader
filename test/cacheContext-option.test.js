const path = require('path');
const normalizePath = require('normalize-path');
const { webpack } = require('./helpers');

const mockCacheLoaderWriteFn = jest.fn();
const mockBaseWebpackConfig = {
  loader: {
    options: {
      write: (cacheKey, cacheData, callback) => {
        mockCacheLoaderWriteFn(cacheKey, cacheData, callback);
        callback(null, ...cacheData.result);
      },
    },
  },
};
const mockRelativeWebpackConfig = {
  loader: {
    options: {
      cacheContext: path.resolve('.'),
      write: (cacheKey, cacheData, callback) => {
        mockCacheLoaderWriteFn(cacheKey, cacheData, callback);
        callback(null, ...cacheData.result);
      },
    },
  },
};

const buildSnapshotReadyDeps = (deps) =>
  deps
    .sort()
    .map((dep) =>
      Object.assign({}, dep, { mtime: null, path: normalizePath(dep.path) })
    );

const buildCacheLoaderCallsData = (calls) =>
  calls.sort().map((rawCall) => {
    const [, call] = rawCall;
    return {
      ...call,
      remainingRequest: normalizePath(call.remainingRequest),
      dependencies: buildSnapshotReadyDeps(call.dependencies),
      contextDependencies: buildSnapshotReadyDeps(call.contextDependencies),
    };
  });

describe('cacheContext option', () => {
  it('should generate relative paths to the project root', async () => {
    const testId = './basic/index.js';
    const stats = await webpack(testId, mockRelativeWebpackConfig);

    const cacheLoaderCallsData = buildCacheLoaderCallsData(
      mockCacheLoaderWriteFn.mock.calls
    );

    expect(
      cacheLoaderCallsData.every(
        (call) => !call.remainingRequest.includes(path.resolve('.'))
      )
    );
    expect(cacheLoaderCallsData).toMatchSnapshot('generated cache-loader data');
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });

  it('should generate absolute paths to the project root', async () => {
    const testId = './basic/index.js';
    const stats = await webpack(testId, mockBaseWebpackConfig);

    const cacheLoaderCallsData = buildCacheLoaderCallsData(
      mockCacheLoaderWriteFn.mock.calls
    );

    expect(
      cacheLoaderCallsData.every((call) =>
        call.remainingRequest.includes(path.resolve('.'))
      )
    );
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });
});
