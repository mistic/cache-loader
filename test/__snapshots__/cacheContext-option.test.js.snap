// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors 1`] = `Array []`;

exports[`errors 2`] = `Array []`;

exports[`generated cache-loader data 1`] = `
Array [
  Object {
    "contextDependencies": Array [],
    "dependencies": Array [
      Object {
        "mtime": null,
        "path": "test/fixtures/basic/file_1.js",
      },
      Object {
        "mtime": null,
        "path": "src/index.js",
      },
    ],
    "remainingRequest": "test/fixtures/basic/file_1.js",
    "result": Array [
      "/* eslint-disable */
console.log('file_1');
",
    ],
  },
  Object {
    "contextDependencies": Array [],
    "dependencies": Array [
      Object {
        "mtime": null,
        "path": "test/fixtures/basic/file_2.js",
      },
      Object {
        "mtime": null,
        "path": "src/index.js",
      },
    ],
    "remainingRequest": "test/fixtures/basic/file_2.js",
    "result": Array [
      "/* eslint-disable */
console.log('file_2');
",
    ],
  },
  Object {
    "contextDependencies": Array [],
    "dependencies": Array [
      Object {
        "mtime": null,
        "path": "test/fixtures/basic/index.js",
      },
      Object {
        "mtime": null,
        "path": "src/index.js",
      },
    ],
    "remainingRequest": "test/fixtures/basic/index.js",
    "result": Array [
      "/* eslint-disable */
require('./file_1.js');
require('./file_2.js');

console.log('basic_entry');
",
    ],
  },
]
`;

exports[`warnings 1`] = `Array []`;

exports[`warnings 2`] = `Array []`;
